File: jvm/src/main/kotlin/com/jetbrains/signatureverifier/serialization/V2CertificateAttributeInfo.kt
package com.jetbrains.signatureverifier.serialization

import kotlinx.serialization.Serializable
import org.bouncycastle.asn1.DLSequence
import org.bouncycastle.asn1.cms.Attribute

@Serializable
data class V2CertificateAttributeInfo(
  override val identifier: TextualInfo,
  val content: EncodableInfo
) : AttributeInfo {

  constructor(attribute: Attribute) : this(
    TextualInfo.getInstance(attribute.attrType),
    attribute.attrValues.toEncodableInfo()
  )

  override fun toAttributeDLSequence(): DLSequence =
    listOf(identifier.toPrimitive(), content.toPrimitive()).toDLSequence()
}

File: net/JetBrains.Serialization/src/Dataholders/V2CertificateAttributeInfo.cs
using Newtonsoft.Json;
using Org.BouncyCastle.Asn1;
using Attribute = Org.BouncyCastle.Asn1.Cms.Attribute;

namespace JetBrains.Serialization;

[JsonObject(MemberSerialization.Fields)]
public class V2CertificateAttributeInfo : AttributeInfo
{
    public override TextualInfo Identifier { get; }
    public IEncodableInfo Content { get; }

    public V2CertificateAttributeInfo(TextualInfo identifier, IEncodableInfo content)
    {
        Identifier = identifier;
        Content = content;
    }

    public V2CertificateAttributeInfo(Attribute attribute)
        : this(TextualInfo.GetInstance(attribute.AttrType), attribute.AttrValues.ToEncodableInfo())
    {
    }

    public override Asn1Encodable GetPrimitiveContent() => Content.ToPrimitive();
}